### ISSUE TITLE
Detecting collapsable if is too eager

### ISSUE THREAD

**Issue by nwr** [Manishearth/rust-clippy#79]
This snippet is warned with `warning: This if statement can be collapsed. Try: if a == 0 && b == 0, #[warn(collapsible_if)] on by default`. Putting a semicolon after the second if block removes the warning, but doesn't seem like it should be necessary.

```rust
    let a = 0;
    let b = 0;
    if a == 0 {
        print!("a is zero");
        if b == 0 {
            println!("a and b is zero");
        }
    }
```

**Comment by Manishearth** [comment]
@llogiq we should do macro-checking here too I guess.


Also, [this](https://github.com/Manishearth/rust-clippy/blob/master/src/collapsible_if.rs#L69) is slightly wrong, it's also okay for the stmts to be 0 but the expr to be some.

**Comment by llogiq** [comment]
I'm currently trying to retrofit the is_empty check to len_zero, but perhaps I can look into this tomorrow.

**Comment by llogiq** [comment]
That said, macro checking won't help us here.

**Comment by llogiq** [comment]
Running the tests, I see that macro checking will be useful anyway (because it will rid us of some spurious warnings). I'll look into why we falsely match on the specific example.

**Comment by killercup** [comment]
Yes, I'd really like it if you could skip macro code, too.

While you fixed `mut_mut` to not trigger on `regex!` stuff, the collapsible-if now generates enough warnings to overflow my terminal… ;)

```bash
src/helpers/normalize.rs:22:32: 22:113 warning: This if statement can be collapsed. Try: if (self.chars.prev.is_some()) && (match c {
    '\t' ...'\r' => true,
    ' ' ...' ' => true,
    '\u{85}' ...'\u{85}' => true,
    '\u{a0}' ...'\u{a0}' => true,
    '\u{1680}' ...'\u{1680}' => true,
    '\u{2000}' ...'\u{200a}' => true,
    '\u{2028}' ...'\u{2029}' => true,
    '\u{202f}' ...'\u{202f}' => true,
    '\u{205f}' ...'\u{205f}' => true,
    '\u{3000}' ...'\u{3000}' => true,
    _ => false,
}), #[warn(collapsible_if)] on by default
src/helpers/normalize.rs:22     let cross_subsection_ref = regex!(r"(?m)^\[(?P<id>.+)\]:\s(?P<file>[^:^/]+)\.html#(?P<subsection>[\w-_]+)$");
                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/main.rs:1:1: 24:1 note: in expansion of regex!
```

As always, you're free to include my code in your tests cases!

**Comment by llogiq** [comment]
I have set up a branch that fixes both macros and block checking. I'm refactoring a bit more, then we can merge.

**Comment by killercup** [comment]
Awesome! You're doing really great stuff here, @llogiq!

**Comment by llogiq** [comment]
Thanks for the kind words.

Next up, I'd like to build something akin to symbolic execution (like in [ASM](http://asm.ow2.org/)). This, combined with a naive SAT checker and a simple form of constant folding can be used to allow for more static reasoning about the code, like finding out incompatible assertions (e.g. `if x == 1 { if x == 2 { ... } }`), tautologies, infinite loops, etc.

**Comment by llogiq** [comment]
Done.
